compiler_gym/service/runtime/BenchmarkCache.cc

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+#include "compiler_gym/service/runtime/BenchmarkCache.h"
+
+#include <glog/logging.h>
+
+using grpc::Status;
+using grpc::StatusCode;
+
+namespace compiler_gym::runtime {
+
+BenchmarkCache::BenchmarkCache(std::optional<std::mt19937_64> rand, size_t maxSizeInBytes)
+    : rand_(rand.has_value() ? *rand : std::mt19937_64(std::random_device()())),
+      maxSizeInBytes_(maxSizeInBytes),
+      sizeInBytes_(0){};
+
+const Benchmark* BenchmarkCache::get(const std::string& uri) const {
```

Comment:
Consider (but not for long, you know what I'm like):
```c++
class BenchmarkCache {
public:
  BenchmarkCache(std::function<Benchmark(string url)> benchmarkCreatorThingie, ...)...
  ...
  Benchmark& operator[] (string url) {
    if (url in cache) return cache[url];
    auto b = benchmarkCreatorThingie(url); // May throw std::out_of_range like std::vector.at
    if (doesn't have capacity) {
      evictUntilCapacity(b.size());
    }
    cache[url] = b;
    return b;
  }
}```

Reply:
I don't know if that makes sense for this use case because there is no `std::function<Benchmark(string url)> benchmarkCreatorThingie` callback that would make sense. If the benchmark can't be found, we return an error to the frontend. Pseudo-code:

```py
class CompilerGymRuntime:
    def session_rpc_endpoint(self, benchmark_uri):
        if benchmark_uri not in self.benchmark_cache:
            return ErrorCode.NOT_FOUND
        benchmark = self.benchmark_cache[benchmark_uri]
        ...
```